optuna/study/_optimize.py

```diff
@@ -237,93 +222,37 @@ def _run_trial(
 
     # `Study.tell` may raise during trial post-processing.
     try:
-        study.tell(trial, values=values, state=state)
+        with warnings.catch_warnings(record=True) as tell_warnings:
+            frozen_trial = study.tell(trial, values=value_or_values, state=state)
     except Exception:
+        frozen_trial = study._storage.get_trial(trial._trial_id)
         raise
     finally:
-        if state == TrialState.COMPLETE:
-            study._log_completed_trial(trial, cast(List[float], values))
-        elif state == TrialState.PRUNED:
-            _logger.info("Trial {} pruned. {}".format(trial.number, str(func_err)))
-        elif state == TrialState.FAIL:
+        if frozen_trial.state == TrialState.COMPLETE:
+            study._log_completed_trial(frozen_trial)
+        elif frozen_trial.state == TrialState.PRUNED:
+            _logger.info("Trial {} pruned. {}".format(frozen_trial.number, str(func_err)))
+        elif frozen_trial.state == TrialState.FAIL:
             if func_err is not None:
                 _logger.warning(
                     "Trial {} failed because of the following error: {}".format(
-                        trial.number, repr(func_err)
+                        frozen_trial.number, repr(func_err)
                     ),
                     exc_info=func_err_fail_exc_info,
                 )
-            elif values_conversion_failure_message is not None:
-                _logger.warning(values_conversion_failure_message)
-            else:
-                assert False, "Should not reach."
+            if tell_warnings is not None:
+                for tell_warning in tell_warnings:
+                    _logger.warning(tell_warning.message)
```

Comment:
This is not a strong request change but more of a verification, but I though that the formats of the "post-trial" log messages were going to be unified in `_run_trial` (the formats of the warnings emitted by `Study.tell` don't necessarily have to be aligned with those). With the current logic they're split between `_run_trial` and in `Study.tell`. Do we share the same understanding?

Reply:
Does the unification you mention looks like following?

```diff
diff --git a/optuna/study/study.py b/optuna/study/study.py
index be175e4b..6d9a4d82 100644
--- a/optuna/study/study.py
+++ b/optuna/study/study.py
@@ -702,6 +702,7 @@ class Study:
 
         if state is None and values is None:
             warnings.warn(
+                f"Trial {frozen_trial.number} failed. "
                 "Study.tell is called with `values` and `state` set to None but "
                 "you have to specify either `values` or `state` at least. "
                 "`values` is required if a trial finishes successfully or "
@@ -730,7 +731,9 @@ class Study:
                 # This path is reached when the objective returns invalid data
                 # including NaN or non-numerical points during Study.optimize.
                 else:
-                    warnings.warn(values_conversion_failure_message)
+                    warnings.warn(
+                        f"Trial {frozen_trial.number} failed. {values_conversion_failure_message}"
+                    )
                     state = TrialState.FAIL
 
             # When called from `Study.optimize` and `state` is pruned, the given `values` contains
```

Reply:
Sorry if my comment was unclear. Not really, what I meant was the following. The trial logs follow a certain convention, i.e. "Trial <some number> .... ", although implicit. At the same time, the logs messages are created (defined in code) at different places with not so good code locality. What I'm suggesting is to improve the code locality. Again, it might be okay to leave it as future work.

Reply:
Thank you so much for explaining your intention in detail. ❤️ It may make sense to me.
So... you think we can create a logging helper for failed trials looks like 9e48748?

Reply:
Thanks a slot, that's more or less exactly what I meant!
I think we can follow this up in a separate ticket about further improving the locality, e.g. make them all Study methods, introducing a new private submodule, or something.